### PR TITLE
Refactoring of asynchronous tests via Promise

### PR DESCRIPTION
**Pull request by LavrovArtem**

### CODE REVIEW

File: test/client/before-test.js [hunk 1]
```diff
@@ -138,6 +138,31 @@
         ok(passed);
     };
 
+    window.createIframeForTest = function (src, elLocation) {
```
Comment (miherlosev) [inline test/client/before-test.js]: `window.createTestIframe (src)`. Now, to simplify let's drop the second parameter.
No sure that we will use this function for nested iframes.
Also let's try this function for several tests

Comment (LavrovArtem) [inline test/client/before-test.js]: In the [test/client/fixtures/utils/dom-test.js](https://github.com/DevExpress/testcafe-hammerhead/pull/1255/commits/07311e417c4e158f3e986b2d29a05a45db5d7b7c#diff-b6a1dd5e49393e1752632fcf8747eb8c) file the createIframeForTest function is used 21 times.
including for nested iframes - https://github.com/DevExpress/testcafe-hammerhead/pull/1255/commits/07311e417c4e158f3e986b2d29a05a45db5d7b7c#diff-b6a1dd5e49393e1752632fcf8747eb8cR682

Comment (miherlosev) [inline test/client/before-test.js]: Ok. Then `window.createTestIframe (src, iframeParentNode)`

Comment (LavrovArtem) [inline test/client/before-test.js]: Maybe ```window.createTestIframe (src, parent)```?

Comment (miherlosev) [inline test/client/before-test.js]: Ok

**Comment by LavrovArtem** [comment]
FPR

File: test/client/fixtures/api-test.js [hunk 2]
```diff
@@ -11,34 +11,30 @@ QUnit.testDone(function () {
...
+        .then(function () {
+            return new hammerhead.Promise(function (resolve) {
+                iframe.addEventListener('load', resolve);
+                setTimeout(resolve, 5000);
+
+                iframe.contentWindow['%hammerhead%'].navigateTo('/test.html');
+            });
+        })
+        .then(function (event) {
+            ok(!event, 'should prevent navigation');
```
Comment (miherlosev) [inline test/client/fixtures/api-test.js]: We don't check the event.
We check the navigation, so change to the origin condition `ok(false, 'should prevent navigation');`

File: test/client/before-test.js [hunk 3]
```diff
@@ -138,6 +147,38 @@
         ok(passed);
     };
 
+    window.createTestIframe = function (attrs, parent) {
+        var iframe = document.createElement('iframe');
+
+        iframe.id = 'test' + Date.now();
+
+        if (typeof attrs === 'string')
```
Comment (miherlosev) [inline test/client/before-test.js]: I propose to drop this and always specify the full notation.
It is more readable.

Compare:
```js
window.createTestIframe('');
window.createTestIframe('about:blank');
window.createTestIframe(window.getSameDomainPageUrl(''))
versus

window.createTestIframe({ src:'' });
window.createTestIframe({ src: 'about:blank' });
window.createTestIframe({ src: window.getSameDomainPageUrl('') })

```

File: test/client/fixtures/babel-transform-test.js [hunk 4]
```diff
@@ -1,14 +1,10 @@
 module('regression');
 
-asyncTest('should not throw an error after for...of loop transformation (GH-1231)', function () {
-    var iframe = document.createElement('iframe');
-    var div    = document.createElement('div');
+test('should not throw an error after for...of loop transformation (GH-1231)', function () {
+    return window.createTestIframe(window.getSameDomainPageUrl('../data/iframe/simple-iframe.html'))
```
Comment (miherlosev) [inline test/client/fixtures/babel-transform-test.js]: Without `window`.
`return createTestIframe(getSameDomainPageUrl('../data/iframe/simple-iframe.html'))`

File: test/client/before-test.js (hunk 3 above)
Comment (LavrovArtem) [inline test/client/before-test.js]: @AlexanderMoskovkin What do you think?

File: test/client/fixtures/babel-transform-test.js (hunk 4 above)
Comment (LavrovArtem) [inline test/client/fixtures/babel-transform-test.js]: eslint:
```
  514:15  error  'getSameDomainPageUrl' is not defined  no-undef
  516:12  error  'createTestIframe' is not defined      no-undef
```

Comment (miherlosev) [inline test/client/fixtures/babel-transform-test.js]: Methods should be added to the [.eslintrc](https://github.com/DevExpress/testcafe-hammerhead/blob/master/test/client/.eslintrc#L20) appropriately.

Comment (LavrovArtem) [inline test/client/fixtures/babel-transform-test.js]: ok

File: test/client/before-test.js (hunk 3 above)
Comment (AlexanderMoskovkin) [inline test/client/before-test.js]: As you wish =) @miherlosev's approach is ok

**Comment by LavrovArtem** [comment]
FPR

**Comment by miherlosev** [comment]
@AlexanderMoskovkin review too. Especially `page-navigation-watch-test.js`

File: test/client/fixtures/api-test.js [hunk 5]
```diff
@@ -11,34 +11,30 @@ QUnit.testDone(function () {
...
+    iframe.id = 'test' + Date.now();
+
+    return hammerhead.Promise.resolve()
+        .then(function () {
+            return new hammerhead.Promise(function (resolve) {
+                iframe.setAttribute('src', 'about:blank');
+                iframe.addEventListener('load', resolve);
+                document.body.appendChild(iframe);
+            });
+        })
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/api-test.js]: why we need `hammerhead.Promise.resolve` ? Why don't start with 
```js
return new hammerhead.Promise(....)
    .then()
```

File: test/client/fixtures/page-navigation-watch-test.js [hunk 6]
```diff
@@ -26,328 +27,209 @@ QUnit.testDone(function () {
...
-        .then(function () {
-            ok(false);
-            start();
-        })
-        .catch(function () {
-            ok(true);
-            start();
+test('Hash via location.href', function () {
+    return navigateIframe('location.href += "#hash";', { timeout: 500 })
+        .catch(function (reason) {
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/page-navigation-watch-test.js]: we should check that the `pageNavigationTriggered` event is not triggered here. The previous version of the test checked it. But the modified don't do it.
Now if the event will be raised the `navigateIframe` promise will be resolved and test will finish succesfully but it should fail.

File: test/client/fixtures/page-navigation-watch-test.js [hunk 7]
```diff
@@ -26,328 +27,209 @@ QUnit.testDone(function () {
...
-            ok(false);
-            start();
-        })
-        .catch(function () {
-            ok(true);
-            start();
+test('Hash via location.hash', function () {
+    return navigateIframe('location.hash = "hash";', { timeout: 500 })
+        .catch(function (reason) {
+            ok(reason === 'timeout exceeded', reason);
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/page-navigation-watch-test.js]: the same as above

File: test/client/fixtures/page-navigation-watch-test.js [hunk 8]
```diff
@@ -358,18 +240,13 @@ asyncTest('Click raised by child node - prevented', function () {
...
-    navigateIframe(navigationScript)
-        .then(function () {
-            ok(false);
-            start();
-        })
-        .catch(function () {
-            ok(true);
-            start();
+    return navigateIframe(navigationScript, { timeout: 500 })
+        .catch(function (reason) {
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/page-navigation-watch-test.js]: the same (here and below)

**Comment by LavrovArtem** [comment]
FPR

**Pull request merged**